smtp_connection_reuse/models/ir_mail_server.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+import threading
+import smtplib
+import logging
+
+from openerp import models
+
+connections_cache = threading.local()
+connections_cache.connections = None
+
+_logger = logging.getLogger(__name__)
+
+
+class SMTPNoQuit(smtplib.SMTP):
+    """ Overrides quit() to prevent Odoo from closing the connection """
+    def quit(self):
+        pass
+
+
+class SMTPSSLNoQuit(smtplib.SMTP_SSL):
+    """ Overrides quit() to prevent Odoo from closing the connection """
+    def quit(self):
+        pass
+
+
+class IrMailServer(models.Model):
+
+    _inherit = 'ir.mail_server'
+
+    def connect(self, host, port, user=None, password=None,
+                encryption=False, smtp_debug=False):
+        if getattr(connections_cache, 'connections', None) is None:
```

Comment:
@andreparames Why `getattr` and not `if not connection_cache.connections`?

Reply:
Each thread has its own attributes, so `connections` might not exist for the current thread. The problematic line is actually the 11th, since it gives the idea it's creating `connections` for all threads, when in fact it only creates it for the current. I'll remove the line.